sympy/tensor/array/ndim_array.py

```diff
@@ -420,9 +426,19 @@ def __eq__(self, other):
         >>> a == b
         False
         """
+        from sympy.tensor.array import SparseNDimArray
         if not isinstance(other, NDimArray):
             return False
-        return (self.shape == other.shape) and (list(self) == list(other))
+
+        if not self.shape == other.shape:
+            return False
+
+        if isinstance(self, SparseNDimArray) and isinstance(other, SparseNDimArray):
+            return self._sparse_array.keys() == other._sparse_array.keys() and \
+                   all(v == other._sparse_array[k] for (k, v) in self._sparse_array.items()
+                       if k in other._sparse_array)
+
+        return list(self) == list(other)
```

Comment:
why not just use something similar to this? dict(...) == dict(... other ...) ?

Reply:
I thought I am not supposed to cast one type to the other so I came up with this code. But if we can cast `Dict` to `dict`, that would indeed be much simpler. :-)
By the way, the problem of comparing `Dict` and `dict` will dispear if the test about `mutableArray.diff() == MutableArray` is changed to `mutableArray.diff() == ImmutableArray()`.

Reply:
Just cast to dict.